### PR TITLE
Add IntelliJ codestyle

### PR DESCRIPTION
**Pull request by MSDehghan**
Add IntelliJ codestyle
+ Disable JavaDocType rule on test files

### CODE REVIEW

File: intellij-codestyle.xml [hunk 1]
```diff
@@ -0,0 +1,598 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<code_scheme name="GoogleStyle">
+  <option name="OTHER_INDENT_OPTIONS">
+    <value>
+      <option name="INDENT_SIZE" value="2" />
+      <option name="CONTINUATION_INDENT_SIZE" value="4" />
+      <option name="TAB_SIZE" value="2" />
```
Comment (bozorgzadeh) [inline intellij-codestyle.xml]:
```suggestion
      <option name="TAB_SIZE" value="4" />
```

File: intellij-codestyle.xml [hunk 2]
```diff
@@ -0,0 +1,598 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<code_scheme name="GoogleStyle">
```
Comment (bozorgzadeh) [inline intellij-codestyle.xml]:
```suggestion
<code_scheme name="Java Code Style from github.com/sahabpardaz/javastyle.guide">
```

File: intellij-codestyle.xml [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+    <value />
+  </option>
+  <option name="IMPORT_LAYOUT_TABLE">
+    <value>
+      <package name="" withSubpackages="true" static="true" />
+      <emptyLine />
+      <package name="" withSubpackages="true" static="false" />
+    </value>
+  </option>
+  <option name="RIGHT_MARGIN" value="100" />
```
Comment (bozorgzadeh) [inline intellij-codestyle.xml]:
```suggestion
  <option name="RIGHT_MARGIN" value="120" />
```

File: README.md [hunk 4]
```diff
@@ -35,4 +35,9 @@ just explain the thrown exceptions. Forcing developers to write summary for all
...
 
 6. **AvoidUnusedImports rules is added:** Although this is not part of the Google Checkstyle,
-it has been added because it cleans the header of Java files by removing misleading references.
\ No newline at end of file
+it has been added because it cleans the header of Java files by removing misleading references.
+
+7. **MissingJavadocType rule is ignored on test files:** Usually test files do not require javadoc as it is obvious 
+what class this test is for.
+
+7. **Suppression file(if exists) must be located beside the main file**.
```
Comment (bozorgzadeh) [inline README.md]: Why you have added this constraint?

File: checkstyle.xml [hunk 5]
```diff
@@ -38,11 +38,16 @@ You can see the original Google java style guide (the base of this customization
         <property name="fileNamePattern" value="module\-info\.java$"/>
     </module>
     <!-- https://checkstyle.org/config_filters.html#SuppressionFilter -->
+    <!-- Customization: If anyone wants to add suppression file, it must be located beside the main file.-->
     <module name="SuppressionFilter">
-        <property name="file" value="${org.checkstyle.google.suppressionfilter.config}"
-            default="checkstyle-suppressions.xml"/>
+        <property name="file" value="${config_loc}/checkstyle-suppressions.xml"/>
         <property name="optional" value="true"/>
     </module>
```
Comment (bozorgzadeh) [inline checkstyle.xml:45]: What is the reason for your change here?

File: README.md [hunk 6]
```diff
@@ -35,4 +35,9 @@ just explain the thrown exceptions. Forcing developers to write summary for all
 from important points to writing a summary to pass the checkstyle.
```
Comment (bozorgzadeh) [inline README.md:46]: Github does not allow me to add comments for non-changed lines on top of this file!!
So I add this comment here: I think we should change the header of the file a little according to its new files. So this is my suggestion for the header of this ReadME:

# Java Style Guide
In Sahab, we follow a code style that is mostly based on [Google code style](https://google.github.io/styleguide/javaguide.html) with a few customizations.  
This Github project contains some config files related to that style:

-  checkstyle.xml: The config file for [Checkstyle](https://checkstyle.sourceforge.io/). Using this file and its associated tools like Maven and Gradle plugins, we can ensure our code style rules in our CI/CD pipelines. To make this file, we have used [the google check style file](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml) as the base file and we have changed it a little according to our customizations.
-  intellij-codestyle.xml: The config file to configure our code style in Intellij. To make this file, we have used [the google Intellij code style file](...) as the base file and we have changed it a little according to our customizations.
-  vscode-codestyle.xml: The config file to configure our code style in VSCode. To make this file, we have used [the google VSCode code style file](...) as the base file and we have changed it a little according to our customizations.

## The customizations 
We used the Google code style after the following customizations:
... [list of the customizations]

File: checkstyle.xml (hunk 5 above)
Comment (MSDehghan) [reply checkstyle.xml:45]: To use suppression file in rare cases (for example to ignore a generated class etc), the user must define a variable named "org.checkstyle.google.suppressionfilter.config" which points to absolute path of suppression file. During my experiments with this feature I found this to be too difficult to use with no real reason.
Beside its difficulty to use, putting checkstyle files in the same place is a best practice and make them more readable. After all I decided to do this change.

File: README.md (hunk 4 above)
Comment (MSDehghan) [reply README.md]: I have explained below

File: README.md (hunk 6 above)
Comment (MSDehghan) [reply README.md:46]: Done

File: README.md [hunk 7]
```diff
@@ -40,4 +36,4 @@ it has been added because it cleans the header of Java files by removing mislead
 7. **MissingJavadocType rule is ignored on test files:** Usually test files do not require javadoc as it is obvious 
 what class this test is for.
 
-7. **Suppression file(if exists) must be located beside the main file**.
\ No newline at end of file
+7. **Suppression file(if exists) must be located beside the main file**.
```
Comment (bozorgzadeh) [inline README.md]: According to the new changes in the structure of this file, this section describes our code style customizations in general, not the exact changes in checkstyle.xml. This item is a minor change in the checkstyle.xml and is not about the changes in rules, so I think it is better not to mention it.

File: README.md [hunk 8]
```diff
@@ -1,18 +1,14 @@
...
 
-## Code Style
-Our code style is based on [Google code style](https://google.github.io/styleguide/javaguide.html) with a
-few customizations. The exact customizations and the reason behind each one of them is described below.
-To check code style on source codes we use [Checkstyle](https://checkstyle.sourceforge.io/). Currently, version 8.45.1
+-  checkstyle.xml: The config file for [Checkstyle](https://checkstyle.sourceforge.io/). Using this file and its associated tools like Maven and Gradle plugins, we can ensure our code style rules in our CI/CD pipelines. To make this file, we have used [the google check style file](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml) as the base file and we have changed it a little according to our customizations. Currently, version 8.45.1
 of Checkstyle is recommended.
+-  intellij-codestyle.xml: The config file to configure our code style in Intellij. To make this file, we have used [the google Intellij code style file](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml) as the base file and we have changed it a little according to our customizations.
```
Comment (bozorgzadeh) [inline README.md]: Please add the code style file of VSCode too. (I think you can get it from Amin or other developers and it is straightforward to do it in this PR)

File: README.md [hunk 9]
```diff
@@ -1,18 +1,14 @@
 # Java Style Guide
-This project contains all style guides used for Java language files.
+In Sahab, we follow a code style that is mostly based on [Google code style](https://google.github.io/styleguide/javaguide.html) with a few customizations.  
+This Github project contains some config files related to that style:
 
-## Code Style
-Our code style is based on [Google code style](https://google.github.io/styleguide/javaguide.html) with a
-few customizations. The exact customizations and the reason behind each one of them is described below.
-To check code style on source codes we use [Checkstyle](https://checkstyle.sourceforge.io/). Currently, version 8.45.1
+-  checkstyle.xml: The config file for [Checkstyle](https://checkstyle.sourceforge.io/). Using this file and its associated tools like Maven and Gradle plugins, we can ensure our code style rules in our CI/CD pipelines. To make this file, we have used [the google check style file](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml) as the base file and we have changed it a little according to our customizations. Currently, version 8.45.1
```
Comment (bozorgzadeh) [inline README.md]: You have used large lines in this item and the next one. Please follow the maximum 120 characters rule.

Comment (MSDehghan) [reply README.md]: Done

**Pull request merged**